Add link rel and link class to image block inspector (see #7504)

greatislander opened the pull request:
## Description

This is some initial work towards addressing #7504. If there's interest from maintainers, I can proceed to add the 'Open link in a new tab' checkbox and investigate automated testing of these features.

## How has this been tested?
- Added link `rel` and link `class` attributes via the block inspector, saved, viewed saved post to ensure that attributes were added.
- Added tests to fixtures.

## Screenshots

<img width="279" alt="Screen shot of image block inspector with link class and link rel fields added." src="https://user-images.githubusercontent.com/605361/42411475-c00c5c98-81d2-11e8-8820-bda14732c21a.png">

## Types of changes
New feature: adds image link `rel` and `class` attributes to core image block.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->

Commit:
Add link classes and rel attribute to image block (see #7504)

Commit:
Make labels consistent with classic editor.

Commit:
Fix coding standards error

Commit:
Add tests

Commit:
Consolidate check for presence of image link

karmatosed reviewed (changes requested):
I would like to know if we need these all, it feels like a lot to add for image. For example do we need custom CSS when we have that on the block?

greatislander commented:
The distinction between the class on an image and the class on a link wrapping the image could certainly make a difference. I think the related issue makes a good point about maintaining existing functionality.

karmatosed commented:
As explored in #7504 I am after considering not totally sure this is a secondary settings, this should be in the actual media library itself:

![artboard](https://user-images.githubusercontent.com/253067/42814921-aa7a93d4-89bd-11e8-8fde-1d907cca2dc7.png)

karmatosed commented:
+1 to @jasmussen's comment to focus us on time.

greatislander commented:
Okay, I'll rebase.

greatislander commented:
Now that https://github.com/WordPress/gutenberg/pull/6316 is merged, should the user-supplied `rel` be additive?

E.g. if I set my image to open a link in a new window and then add a custom `rel` value, should the `rel` be set to `noreferrer noopener myrelvalue`?

getsource commented:
Just a quick note that I've got a functioning version of this I can put in a separate branch if it's wanted. Was just figuring out how to best get it back into Github today.

getsource commented:
It's referenced above, but I pushed the updated changes to this branch:
https://github.com/WordPress/gutenberg/tree/add/image-details

I currently only have the "Open in new tab" adding to `rel` if `rel` does not have content.

Commit:
Rebase, incorporate upstream changes

Commit:
Merge branch 'upstream' into add/image-details

greatislander commented:
Rebased including a couple of tweaks from @getsource that I missed.

getsource commented:
Thanks so much @greatislander!

Tested and seems to be working as expected.

This was passing locally, so I went ahead and re-ran the failed travis test, and it passed.

greatislander commented:
I've followed the discussion in Slack and on #7504 around this; are there still concerns about the link `class` and `rel` sidebar functionality in this PR, or is it sufficient to address those two pieces of #7504?

getsource commented:
@greatislander I believe I just misread @karmatosed comment on #7504, based on https://github.com/WordPress/gutenberg/pull/7771#issuecomment-434029933.

@karmatosed, if that's the case, would you mind removing the change request? If not, we can definitely keep chatting about alternatives.

Commit:
Merge branch 'master' into add/image-details

greatislander commented:
Rebased again.

getsource commented:
Just a quick note that I re-ran it, and it passed.

talldan reviewed (changes requested):
Hi @greatislander, Thanks for working on this 😄  

I've given this a code review, though I think the design questions should probably be answered first before taking this further.

My opinion on the current functionality in this PR is that this offering the ability to freely edit these attributes takes us quite close to being an html editor, whereas Gutenberg should perhaps be offering a higher level of abstraction over these attributes.

I believe @karmatosed expressed when Link Target was added that the sidebar for the image block is getting quite busy, and this does exacerbate that problem.

Shifting this functionality to the media library could be a way to handle that, though I know the integration is quite hard to work with.

I could also imagine checkboxes for popular rel values being an interesting way to represent that to the end user (but I'm not a designer).

greatislander commented:
Hi @talldan, thanks for the review. In terms of the bigger conceptual question of having the link `class` and `rel` in the media library, I don’t think that works; they are applied to the specific link wrapped around an image tag in its particular placement in the editor; the media library allows attributes for the image itself to be edited globally, if I’m not mistaken? I think this was discussed further in Slack, and a separate issue was created to identify parts of #7504 that should be addressed within the media library (as opposed to these two, which are only relevant in the context of a specific linked image in the editor). I’m not sure if the thinking there has changed?

I’m happy work on some of your suggested code changes.

getsource commented:
Here's a GIF of the functionality after the rebase, since it's a little different than before. Notably, the fields are only visible when "link to" is turned on.
![link class and rel](https://user-images.githubusercontent.com/1034160/48229061-d4694d00-e35b-11e8-9fda-8697c4a242d1.gif)

I agree that the media library probably isn't the best place for these settings, since they are unique to the instance of the image that is inserted into the post, and not to the image across the entire WordPress install.

Commit:
Changes as per review

Commit:
Merge branch 'upstream' into add/image-details

Commit:
Fix tests

greatislander commented on packages/block-library/src/image/index.js:
```diff
@@ -90,7 +102,7 @@ const schema = {
 			...imageSchema,
 			a: {
-				attributes: [ 'href', 'target' ],
+				attributes: [ 'linkClass', 'href', 'rel', 'target' ],
```
@talldan Just to be clear, would that mean that `linkClass` would be disregarded if an image wrapped in a link was parsed in the context of a raw transform? I wouldn't think that would be desirable. But maybe I'm misunderstanding.

talldan replied:
@greatislander Seems like linkClass is disregarded anyway. The attributes listed in this schema are not block attributes, they're HTML attributes. They represent a whitelist of attributes that are kept when pasted raw html is parsed.

That parsing happens in a function called `pasteHandler`:
https://github.com/WordPress/gutenberg/blob/baf00986b5624ae88949feadd2b28cd5c94fab72/packages/blocks/src/api/raw-handling/index.js#L126

Eventually, that calls through to cleanNodeList, which uses the matching schema from the block definition and strips any excess attributes:
https://github.com/WordPress/gutenberg/blob/baf00986b5624ae88949feadd2b28cd5c94fab72/packages/blocks/src/api/raw-handling/utils.js#L152

In that function there's special handling for classes on line 177. In the schema classes need to be whitelisted separately (by making them match either a string or a regex) which you can see in the image block:
https://github.com/WordPress/gutenberg/blob/master/packages/block-library/src/image/index.js#L82

I'm not sure what the best practice is here, but it seems like you could define a regex that accepts any valid class, and that would make linkClass start working again.

greatislander commented:
@talldan @karmatosed I've updated this PR with @talldan's suggestions, rebased… any chance of another review? See also @getsource's comment: https://github.com/WordPress/gutenberg/pull/7771#issuecomment-437166143

Commit:
Merge branch 'upstream' into add/image-details

karmatosed reviewed (approved):
Let's get this in for now, we can iterate and review the media library itself in phase 2.

Commit:
Merge branch 'upstream' into add/image-details

Commit:
Resolve final review comment

greatislander commented:
Thanks @karmatosed! Should be merge-ready once tests pass.

Commit:
Update CONTRIBUTORS.md

greatislander commented:
@talldan I've removed the attribute as per your comments— should be ready for another review and merge.

greatislander commented:
@karmatosed @talldan @jasmussen Any chance this can be milestoned for 4.5 and merged? I've resolved all review comments. Thanks!

jasmussen commented:
Yep, sorry about that.

talldan commented:
Hey @greatislander - Thanks for addressing all those comments, and sorry for the delay responding. This does look good, I gave it a test and it works really nicely. 

I don't think the issue with the linkClass working in raw transforms was ever addressed. It'd be great if that could be created as a separate issue or PR.

greatislander commented:
Hi @talldan, thanks -- I will create a separate issue or PR for the linkClass issue.

greatislander commented:
Follow-up ticket: https://github.com/WordPress/gutenberg/issues/11995

Commit:
Merge branch 'upstream' into add/image-details

youknowriad commented on packages/block-library/src/image/index.js:
```diff
@@ -231,7 +269,16 @@ export const settings = {
+						rel={ rel }
+					>
+						{ image }
+					</a>
```
I wonder if this change creates an "invalid image" warning if you try to upgrade from an image with a link set to open a new tab.

 * checkout master
 * add an image with an external link
 * save
 * checkout this branch
 * refresh

The image shouldn't show up as invalid. I wonder if we should add a deprecated version for this?

To clarify: I didn't have the time to test it but it's a guess.

jasmussen replied:
I took this for a spin, and things look like they are working as intended.

Screenshot from master:

<img width="952" alt="screenshot 2018-11-19 at 11 32 48" src="https://user-images.githubusercontent.com/1204802/48701516-fb8f0c80-ebee-11e8-9b5e-e043125b173f.png">

Checked out this branch and reloaded:

<img width="952" alt="screenshot 2018-11-19 at 11 33 05" src="https://user-images.githubusercontent.com/1204802/48701520-0053c080-ebef-11e8-888b-c0525e67b819.png">

I suspect the problem only occurs when you go the other way. Which is adding the attributes, then checking out a branch that removes them. And sure enough:

<img width="949" alt="screenshot 2018-11-19 at 11 34 52" src="https://user-images.githubusercontent.com/1204802/48701623-37c26d00-ebef-11e8-8300-d0d7eca3b840.png">

Going back to master:

<img width="948" alt="screenshot 2018-11-19 at 11 35 16" src="https://user-images.githubusercontent.com/1204802/48701626-38f39a00-ebef-11e8-9fc6-d98fb0d51f07.png">

But this shouldn't be a problem unless we choose to remove this feature :D

jasmussen replied:
Checked again, this time I checkout master first, and set the open externally and added a custom link:

<img width="952" alt="screenshot 2018-11-19 at 11 41 03" src="https://user-images.githubusercontent.com/1204802/48702037-275ec200-ebf0-11e8-8079-f35b08d40bd2.png">

Checking out this branch and reloading still worked as expected:

<img width="954" alt="screenshot 2018-11-19 at 11 41 35" src="https://user-images.githubusercontent.com/1204802/48702050-2fb6fd00-ebf0-11e8-8dbe-4c3d0e0a2b94.png">

👍 👍

youknowriad replied:
Thanks for the tests @jasmussen :)

youknowriad reviewed (approved):
LGTM 👍

StaggerLeee commented:
Don't forget new filter to make desired "Link to" default for all new images.
